In [1]:
# Imports
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
# Imports for further analysis
import matplotlib.pyplot as plt
import seaborn as sns

# Prepare the data to be used on a neural network model
Step 1: Read the applicants_data.csv file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.¶

In [2]:
# Upload stock analysis_lr  to Colab
from google.colab import files
csv_file = files.upload()

Saving stock_analysis_data_lr.csv to stock_analysis_data_lr.csv


In [3]:
# Read the applicants_data.csv file from the Resources folder into a Pandas DataFrame
stock_analysis_data_lr = pd.read_csv('stock_analysis_data_lr.csv')

stock_analysis_data_lr.drop(columns=['date'],inplace=True)
# Review the DataFrame
# YOUR CODE HERE
stock_analysis_data_lr


EPS  treausry_yield  inflation_expectation  sp500_return  \
0     3.84        0.090909              -0.035714      0.022148   
1     3.84        0.090909              -0.035714      0.022148   
2     3.84        0.090909              -0.035714      0.022148   
3     3.84        0.090909              -0.035714      0.022148   
4     3.84        0.090909              -0.035714      0.022148   
...    ...             ...                    ...           ...   
3984  3.35        0.175325               0.000000      0.037556   
3985  3.35        0.175325               0.000000      0.037556   
3986  3.35        0.175325               0.000000      0.037556   
3987  3.35        0.175325               0.000000      0.037556   
3988  3.35        0.175325               0.000000      0.037556   

      unemployment_rate  consumer_sentiment  %-close_AAPL-close  \
0             -0.053571            0.059537                 1.0   
1             -0.053571            0.059537                 0.0   
2             -0.053571            0.059537                 0.0   
3             -0.053571            0.059537                 0.0   
4             -0.053571            0.059537                 0.0   
...                 ...                 ...                 ...   
3984          -0.050000           -0.065476                 0.0   
3985          -0.050000           -0.065476                 0.0   
3986          -0.050000           -0.065476                 0.0   
3987          -0.050000           -0.065476                 0.0   
3988          -0.050000           -0.065476                 0.0   

      %-close_ABBV-close  %-close_ABT-close  %-close_AMGN-close  ...  \
0                    0.0                0.0                 0.0  ...   
1                    0.0                0.0                 0.0  ...   
2                    0.0                0.0                 0.0  ...   
3                    0.0                0.0                 0.0  ...   
4                    0.0                0.0                 0.0  ...   
...                  ...                ...                 ...  ...   
3984                 0.0                0.0                 0.0  ...   
3985                 0.0                0.0                 0.0  ...   
3986                 0.0                0.0                 0.0  ...   
3987                 0.0                0.0                 0.0  ...   
3988                 0.0                1.0                 0.0  ...   

      %-close_T-close  %-close_UNH-close  %-close_UTX-close  %-close_V-close  \
0                 0.0                0.0                0.0              0.0   
1                 0.0                0.0                0.0              0.0   
2                 0.0                0.0                0.0              0.0   
3                 0.0                0.0                0.0              0.0   
4                 0.0                0.0                0.0              0.0   
...               ...                ...                ...              ...   
3984              0.0                0.0                0.0              0.0   
3985              0.0                0.0                0.0              0.0   
3986              0.0                0.0                1.0              0.0   
3987              0.0                0.0                0.0              0.0   
3988              0.0                0.0                0.0              0.0   

      %-close_VZ-close  %-close_WFC-close  %-close_WMT-close  \
0                  0.0                0.0                0.0   
1                  0.0                0.0                0.0   
2                  0.0                0.0                0.0   
3                  0.0                0.0                0.0   
4                  0.0                0.0                0.0   
...                ...                ...                ...   
3984               0.0                0.0                0.0   
3985               0.0                0.0                0.

In [4]:
# Review the data types associated with the columns
# YOUR CODE HERE
stock_analysis_data_lr.dtypes

EPS                      float64
treausry_yield           float64
inflation_expectation    float64
sp500_return             float64
unemployment_rate        float64
consumer_sentiment       float64
%-close_AAPL-close       float64
%-close_ABBV-close       float64
%-close_ABT-close        float64
%-close_AMGN-close       float64
%-close_AMZN-close       float64
%-close_BA-close         float64
%-close_BAC-close        float64
%-close_BMY-close        float64
%-close_BRK.B-close      float64
%-close_C-close          float64
%-close_CMCSA-close      float64
%-close_CSCO-close       float64
%-close_CVX-close        float64
%-close_DIS-close        float64
%-close_DWDP-close       float64
%-close_FB-close         float64
%-close_GE-close         float64
%-close_GILD-close       float64
%-close_GOOG-close       float64
%-close_GOOGL-close      float64
%-close_HD-close         float64
%-close_HON-close        float64
%-close_IBM-close        float64
%-close_INTC-close       float64
%-close_JN

# Step 2: Using the preprocessed data, create the features (X) and target (y) datasets. The target dataset should be defined by the preprocessed DataFrame column “TARGET”. The remaining columns should define the features dataset

In [5]:
#Create Features lits for input to the analysis
FEATURES = list(stock_analysis_data_lr.iloc[:,:-1])
FEATURES

['EPS',
 'treausry_yield',
 'inflation_expectation',
 'sp500_return',
 'unemployment_rate',
 'consumer_sentiment',
 '%-close_AAPL-close',
 '%-close_ABBV-close',
 '%-close_ABT-close',
 '%-close_AMGN-close',
 '%-close_AMZN-close',
 '%-close_BA-close',
 '%-close_BAC-close',
 '%-close_BMY-close',
 '%-close_BRK.B-close',
 '%-close_C-close',
 '%-close_CMCSA-close',
 '%-close_CSCO-close',
 '%-close_CVX-close',
 '%-close_DIS-close',
 '%-close_DWDP-close',
 '%-close_FB-close',
 '%-close_GE-close',
 '%-close_GILD-close',
 '%-close_GOOG-close',
 '%-close_GOOGL-close',
 '%-close_HD-close',
 '%-close_HON-close',
 '%-close_IBM-close',
 '%-close_INTC-close',
 '%-close_JNJ-close',
 '%-close_JPM-close',
 '%-close_KO-close',
 '%-close_MA-close',
 '%-close_MCD-close',
 '%-close_MDT-close',
 '%-close_MMM-close',
 '%-close_MO-close',
 '%-close_MRK-close',
 '%-close_MSFT-close',
 '%-close_NFLX-close',
 '%-close_NKE-close',
 '%-close_NVDA-close',
 '%-close_ORCL-close',
 '%-close_PEP-close',
 '%-close_PFE-clo

In [6]:
#Prepare the test and train data
test_frac = 0.2
Xlr = stock_analysis_data_lr[FEATURES]
ylr = stock_analysis_data_lr['value']
x_train_lr, x_test_lr, y_train_lr, y_test_lr = train_test_split(Xlr,ylr, test_size = test_frac)
x_train_lr

EPS  treausry_yield  inflation_expectation  sp500_return  \
2909  1.84       -0.338983              -0.045455      0.047645   
1225  2.97        0.056497               0.000000      0.002918   
1743  2.45        0.044444               0.037037      0.005755   
3911  3.35        0.252033               0.020833     -0.029517   
3635  2.33       -0.095238               0.119048      0.029760   
...    ...             ...                    ...           ...   
2572  2.68       -0.025478              -0.107143      0.036198   
3200  2.58        0.259259               0.000000      0.108777   
3808  2.52        0.081081               0.020833      0.046288   
295   4.84        0.017964              -0.037037     -0.049787   
3862  3.35        0.025000              -0.020408     -0.052741   

      unemployment_rate  consumer_sentiment  %-close_AAPL-close  \
2909           2.340909           -0.194164                 0.0   
1225           0.000000           -0.016842                 1.0   
1743          -0.050000           -0.008097                 0.0   
3911           0.025641           -0.048159                 0.0   
3635          -0.084746           -0.050292                 0.0   
...                 ...                 ...                 ...   
2572           0.028571            0.024678                 0.0   
3200          -0.126582            0.017413                 0.0   
3808          -0.086957           -0.059972                 0.0   
295           -0.019608            0.014239                 0.0   
3862          -0.071429            0.047478                 0.0   

      %-close_ABBV-close  %-close_ABT-close  %-close_AMGN-close  ...  \
2909                 0.0                0.0                 0.0  ...   
1225                 0.0                0.0                 0.0  ...   
1743                 0.0                0.0                 0.0  ...   
3911                 0.0                0.0                 0.0  ...   
3635                 0.0                0.0                 0.0  ...   
...                  ...                ...                 ...  ...   
2572                 0.0                0.0                 0.0  ...   
3200                 0.0                0.0                 0.0  ...   
3808                 0.0                0.0                 0.0  ...   
295                  0.0                0.0                 0.0  ...   
3862                 0.0                0.0                 0.0  ...   

      %-close_PM-close  %-close_T-close  %-close_UNH-close  %-close_UTX-close  \
2909               0.0              0.0                0.0                0.0   
1225               0.0              0.0                0.0                0.0   
1743               0.0              0.0                0.0                0.0   
3911               0.0              0.0                0.0                0.0   
3635               0.0              0.0                0.0                0.0   
...                ...              ...                ...                ...   
2572               0.0              0.0                0.0                0.0   
3200               0.0              0.0                0.0                0.0   
3808               0.0              0.0                0.0                0.0   
295                0.0              0.0                0.0                0.0   
3862               0.0              0.0                0.0                0.0   

      %-close_V-close  %-close_VZ-close  %-close_WFC-close  %-close_WMT-close  \
2909              0.0               0.0                0.0                0.0   
1225              0.0               0.0                0.0                0.0   
1743              0.0               0.0                0.0                0.0   
3911              0.0               1.0                0.0                0.0   
3635              0.0               0.0                0.0                0.0   
...               ...               ...                ...               

# Compile and Evaluate a Linear Regression Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

In [7]:
# Define the the number of inputs (features) to the model
number_input_features = 57

# Review the number of features
number_input_features


57

In [8]:
# Define the number of neurons in the output layer
number_output_neurons = 1
number_output_neurons

1

In [9]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 = round((number_input_features+number_output_neurons)/2)
# Review the number hidden nodes in the first layer
hidden_nodes_layer1

29

In [10]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 =  round((hidden_nodes_layer1+number_output_neurons)/2)
# Review the number hidden nodes in the second layer
hidden_nodes_layer2

15

In [11]:
# Create the Sequential model instance
nn = Sequential()

In [12]:
# Add the first hidden layer
# YOUR CODE HERE
nn.add(Dense(
    units = hidden_nodes_layer1,
    activation = 'relu',
    input_dim = number_input_features
))

In [13]:
# Add the second hidden layer
# YOUR CODE HERE
nn.add(Dense(
    units = hidden_nodes_layer1,
    activation = 'relu',
))

In [14]:
# Add the third hidden layer
# YOUR CODE HERE
nn.add(Dense(
    units = hidden_nodes_layer2,
    activation = 'relu'
))

In [15]:
# Add the fourth hidden layer
# YOUR CODE HERE
nn.add(Dense(
    units = hidden_nodes_layer2,
    activation = 'relu'
))

In [16]:
# Add the output layer to the model specifying the number of output neurons and activation function
# YOUR CODE HERE
nn.add(Dense(units=1, activation ='linear'))

In [17]:
# Display the Sequential model summary
# YOUR CODE HERE
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 29)                1682      
                                                                 
 dense_1 (Dense)             (None, 29)                870       
                                                                 
 dense_2 (Dense)             (None, 15)                450       
                                                                 
 dense_3 (Dense)             (None, 15)                240       
                                                                 
 dense_4 (Dense)             (None, 1)                 16        
                                                                 
Total params: 3,258
Trainable params: 3,258
Non-trainable params: 0
_________________________________________________________________


### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.

In [18]:
# Compile the Sequential model
# YOUR CODE HERE
nn.compile(optimizer='adam',loss='mse',metrics='mean_absolute_error')

In [19]:
# Fit the model using 50 epochs and the training data
# YOUR CODE HERE
nn.fit(x_train_lr, y_train_lr, epochs=500)

Epoch 1/500
100/100 [==============================] - 1s 2ms/step - loss: 190000.0625 - mean_absolute_error: 179.0871
Epoch 2/500
100/100 [==============================] - 0s 2ms/step - loss: 162843.6719 - mean_absolute_error: 159.8304
Epoch 3/500
100/100 [==============================] - 0s 2ms/step - loss: 153678.3438 - mean_absolute_error: 178.8436
Epoch 4/500
100/100 [==============================] - 0s 2ms/step - loss: 145729.6406 - mean_absolute_error: 167.3360
Epoch 5/500
100/100 [==============================] - 0s 2ms/step - loss: 128494.1172 - mean_absolute_error: 156.7522
Epoch 6/500
100/100 [==============================] - 0s 2ms/step - loss: 90758.6797 - mean_absolute_error: 116.1854
Epoch 7/500
100/100 [==============================] - 0s 2ms/step - loss: 53112.3125 - mean_absolute_error: 86.8635
Epoch 8/500
100/100 [==============================] - 0s 2ms/step - loss: 41474.6680 - mean_absolute_error: 72.5432
Epoch 9/500
100/100 [==============================] 

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.

In [20]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
mean_absolute_error = nn.evaluate(x_test_lr,y_test_lr,verbose=2)

# Display the model loss and accuracy results
print(f"MSE: {mean_absolute_error }")

25/25 - 0s - loss: 19997.3828 - mean_absolute_error: 50.5380 - 159ms/epoch - 6ms/step
MSE: [19997.3828125, 50.538002014160156]


## Results Analysis: The model provides a mean absolute error of 39.64 that is pretty low considering that we are analyzing the Fortune 50 companies here. 

### Step 4: Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`. 

In [21]:
# Set the model's file path
file_path = ("Resources/nn_stocks1.h5")

# Export your model to a HDF5 file
# YOUR CODE HERE
nn.save(file_path)

# Predict using the NN Model

# Load the file (Supposedly obtained from the Investor) from csv format


In [27]:
# Load the regression test file
from google.colab import files
csv_file = files.upload()

Saving stock_test_regression.csv to stock_test_regression (1).csv


In [28]:
# Read the datar received from investors from the Resources folder into a Pandas DataFrame
stock_test = pd.read_csv('stock_test_regression.csv', index_col = 'date', header =1)
# Review the DataFrame
# YOUR CODE HERE
stock_test

EPS  treausry_yield  inflation_expectation  sp500_return  \
date                                                                 
2022-02  3.35        0.175325                    0.0      0.037556   
2022-02  3.35        0.175325                    0.0      0.037556   
2022-02  3.35        0.175325                    0.0      0.037556   
2022-02  3.35        0.175325                    0.0      0.037556   
2022-02  3.35        0.175325                    0.0      0.037556   

         unemployment_rate  consumer_sentiment  %-close_AAPL-close  \
date                                                                 
2022-02              -0.05           -0.065476                 0.0   
2022-02              -0.05           -0.065476                 0.0   
2022-02              -0.05           -0.065476                 0.0   
2022-02              -0.05           -0.065476                 0.0   
2022-02              -0.05           -0.065476                 0.0   

         %-close_ABBV-close  %-close_ABT-close  %-close_AMGN-close  ...  \
date                                                                ...   
2022-02                 0.0                0.0                 0.0  ...   
2022-02                 0.0                0.0                 0.0  ...   
2022-02                 0.0                0.0                 0.0  ...   
2022-02                 0.0                0.0                 0.0  ...   
2022-02                 0.0                1.0                 0.0  ...   

         %-close_PM-close  %-close_T-close  %-close_UNH-close  \
date                                                            
2022-02               0.0              0.0                0.0   
2022-02               0.0              0.0                0.0   
2022-02               0.0              0.0                0.0   
2022-02               0.0              0.0                0.0   
2022-02               0.0              0.0                0.0   

         %-close_UTX-close  %-close_V-close  %-close_VZ-close  \
date                                                            
2022-02                0.0              0.0               0.0   
2022-02                0.0              0.0               0.0   
2022-02                1.0              0.0               0.0   
2022-02                0.0              0.0               0.0   
2022-02                0.0              0.0               0.0   

         %-close_WFC-close  %-close_WMT-close  %-close_XOM-close  Target  
date                                                                      
2022-02                0.0                0.0                0.0     0.0  
2022-02                0.0                0.0                0.0     0.0  
2022-02                0.0                0.0                0.0     1.0  
2022-02                0.0                0.0                0.0     1.0  
2022-02                0.0                0.0                0.0     0.0  

[5 rows x 57 columns]

In [29]:
# Predict the stock values for the data obtained from the investor
nn_values = nn.predict(stock_test)
nn_values

array([[ 63.20001 ],
       [114.314835],
       [ 69.89576 ],
       [ 56.191475],
       [ 87.24145 ]], dtype=float32)

**Final Analysis**
The values predicted by Neural Networks are 

array([[ 63.20001 ],
       [114.314835],
       [ 69.89576 ],
       [ 56.191475],
       [ 87.24145 ]], dtype=float32)


The Values Predicted by Random Forest Regression are as below

array([ 59.9283  , 126.619401,  64.600623,  64.587928, 112.40737 ])


The actual values are 

value 59.6512 136.2333 102.7 68.166 120.1474


**On a visual examination of the resultsmit is apparent that the Random Forest Regression are more close to the actual values and we will go with that.**